File: capstone-projects/cafe-management-program/csharp/Project_Nhom04/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.ConstrainedExecution;
using System.Text;
using System.Threading.Tasks;

namespace Project_Nhom04
{
    internal class Service : IComparable
    {
        //fields
        protected string sID = "SV";
        protected string sName;
        protected string sType;
        protected int iAmount;
        protected double dPrice;

        //properties
        public string ID { get { return this.sID; } set { this.sID = value; } }
        public string Name { get { return this.sName; } set { this.sName = value; } }
        public string Type { get { return this.sType; } set { this.sType = value; } }
        public int Amount { get { return this.iAmount; } set { this.iAmount = value; } }
        public double Price { get { return this.dPrice; } set { this.dPrice = value; } }

        //contructor
        public Service() { }
        public Service(string id, string name, string type, int amount, double price)
        {
            this.sID = id;
            this.sName = name;
            this.sType = type;
            this.iAmount = amount;
            this.dPrice = price;
        }

        //destructor
        ~Service() { }

        //Input
        public void InputName()
        {
            Console.Write("Service Name: ");
            this.sName = Program.InputString();
        }

        public void InputType()
        {
            Console.WriteLine("Type Of Service ([0] Food - [1] Drink): ");
            int tp = Program.InputNumber(0, 1);
            switch(tp)
            {
                case 0:
                    this.sType = "Food";
                    break;
                case 1:
                    this.sType = "Drink";
                    break;
            }
        }

        public void InputAmount()
        {
            Console.Write("Amount: ");
            this.iAmount = Program.InputNumber_Int();
        }

        public void InputPrice()
        {
            Console.Write("Price (VND): ");
            this.dPrice = Program.InputNumber_Double();
        }

        public void Input()
        {
            if (Cafe.lservices.Count() == 0)
                this.sID += (Cafe.lservices.Count()).ToString();
            else
                this.sID += (Convert.ToInt32(FindMaxServiceID() + 1)).ToString();
            Console.WriteLine("Current Service ID: " + this.sID);
            InputName();
            InputType();
            InputAmount();
            InputPrice();
        }

        static public int FindMaxServiceID()
        {
            int id = 0;
            for (int i = 0; i < Cafe.lservices.Count(); i++)
            {
                if (Convert.ToInt32(Cafe.lservices[i].ID.Substring(2)) > id)
                    id = Convert.ToInt32(Cafe.lservices[i].ID.Substring(2));
            }
            return id;
        }

        public void Output()
        {
            Console.WriteLine("\t" + this.sID.PadRight(PadRightMax()) + this.sName.PadRight(PadRightMax()) + this.sType.PadRight(PadRightMax()) + this.iAmount.ToString().PadRight(PadRightMax()) + this.Price);
        }

        static public void OutputFields()
        {
            Console.WriteLine("\n\t" + "[ID]".PadRight(PadRightMax()) + "[NAME]".PadRight(PadRightMax()) + "[TYPE]".PadRight(PadRightMax()) + "[AMOUNT]".PadRight(PadRightMax()) + "[PRICE]");
        }

        //Methods
        public int FindPadRight()
        {
            int max = 1;
            while(this.sID.Length > max
                || this.sName.Length > max
                || this.sType.Length > max
                || this.iAmount.ToString().Length > max
                || this.dPrice.ToString().Length > max)
            {
                max++;
            }
            return max;
        }

        static public int PadRightMax()
        {
            int len = Cafe.lservices[0].FindPadRight();
            for (int i = 1; i < Cafe.lservices.Count(); i++)
            {
                if (Cafe.lservices[i].FindPadRight() > len)
                    len = Cafe.lservices[i].FindPadRight();
            }
            return len + 10;
        }

        public int CompareTo(object obj)
        {
            if (!(obj is Service))
            {
                throw new ArgumentException("Compared Object is not of Service");
            }
            Service sv = obj as Service;
            return this.Type.CompareTo(sv.Type);
        }

        //Sort
        static public void SortID()
        {
            for (int i = 0; i < Cafe.lservices.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lservices.Count(); j++)
                {
                    if (String.Compare(Cafe.lservices[i].ID, Cafe.lservices[j].ID) > 0)
                    {
                        Service tmp = Cafe.lservices[i];
                        Cafe.lservices[i] = Cafe.lservices[j];
                        Cafe.lservices[j] = tmp;
                    }
                }
            }
        }

        static public void SortName()
        {
            for (int i = 0; i < Cafe.lservices.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lservices.Count(); j++)
                {
                    if (String.Compare(Cafe.lservices[i].Name, Cafe.lservices[j].Name) > 0)
                    {
                        Service tmp = Cafe.lservices[i];
                        Cafe.lservices[i] = Cafe.lservices[j];
                        Cafe.lservices[j] = tmp;
                    }
                }
            }
        }

        static public void SortAmount()
        {
            for (int i = 0; i < Cafe.lservices.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lservices.Count(); j++)
                {
                    if (Cafe.lservices[i].Amount > Cafe.lservices[j].Amount)
                    {
                        Service tmp = Cafe.lservices[i];
                        Cafe.lservices[i] = Cafe.lservices[j];
                        Cafe.lservices[j] = tmp;
                    }
                }
            }
        }

        static public void SortPrice()
        {
            for (int i = 0; i < Cafe.lservices.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lservices.Count(); j++)
                {
                    if (Cafe.lservices[i].Price > Cafe.lservices[j].Price)
                    {
                        Service tmp = Cafe.lservices[i];
                        Cafe.lservices[i] = Cafe.lservices[j];
                        Cafe.lservices[j] = tmp;
                    }
                }
            }
        }

        //Login_Methods
        static public int SearchService(string id)
        {
            id = id.ToLower();
            for (int i = 0; i < Cafe.lservices.Count(); i++)
            {
                if (Cafe.lservices[i].ID.ToLower() == id)
                {
                    return i;
                }
            }
            return -1;
        }

        static public void AddNewService()
        {
            Console.WriteLine("\n\t[ADDING NEW SERVICE]\n");
            Service sv = new Service();
            sv.Input();
            Cafe.lservices.Add(sv);
        }

        static public int SelectService()
        {
            string id;
            int pos;
            do
            {
                try
                {
                    Console.Write(" => Now Enter Service ID: ");
                    id = Console.ReadLine();
                    pos = SearchService(id);
                    if (pos == -1)
                        continue;
                    else
                        break;
                }
                catch
                {
                    continue;
                }
            } while (true);
            return pos;
        }

        static public void ResetServiceID()
        {
            for (int i = 0; i < Cafe.lservices.Count(); i++)
            {
                Cafe.lservices[i].ID = "SV" + i.ToString();
            }
        }

        //FILES
        static public void WriteDataService()
        {
            string[] wt = new string[Cafe.lservices.Count()];
            for (int i = 0; i < Cafe.lservices.Count(); i++)
            {
                wt[i] = Cafe.lservices[i].ID + ";" + Cafe.lservices[i].Name + ";" + Cafe.lservices[i].Type + ";" + Cafe.lservices[i].Amount.ToString() + ";" + Cafe.lservices[i].Price.ToString();
            }

            File.WriteAllLines("Service.txt", wt);
        }

        static public void ReadDataService()
        {
            string[] a = File.ReadAllLines("Service.txt");

            for (int i = 0; i < a.Length; i++)
            {
                string[] b = a[i].Split(';');
                Service sv = new Service(b[0], b[1], b[2], Convert.ToInt32(b[3]), Convert.ToDouble(b[4]));
                Cafe.lservices.Add(sv);
            }
        }
    }
}

File: capstone-projects/cafe-management-program/java/src/Service.java
import java.io.*;
import java.util.ArrayList;
import java.util.Scanner;

public class Service {
    //fields
    protected String sID = "SV";
    protected String sName;
    protected String sType;
    protected int iAmount;
    protected double dPrice;

    //properties
    public String getsID() {
        return sID;
    }

    public void setsID(String sID) {
        this.sID = sID;
    }

    public String getsName() {
        return sName;
    }

    public String getsType() {
        return sType;
    }

    public int getiAmount() {
        return iAmount;
    }

    public void setiAmount(int iAmount) {
        this.iAmount = iAmount;
    }

    public double getdPrice() {
        return dPrice;
    }

    //constructor
    public Service() {
    }

    public Service(String sID, String sName, String sType, int iAmount, double dPrice) {
        this.sID = sID;
        this.sName = sName;
        this.sType = sType;
        this.iAmount = iAmount;
        this.dPrice = dPrice;
    }

    //methods
    //Input
    public void InputName() {
        System.out.print("Service Name: ");
        this.sName = Main.InputString();
    }

    public void InputType() {
        System.out.println("Type Of Service ([0] Food - [1] Drink): ");
        int tp = Main.InputNumber(0, 1);
        switch (tp) {
            case 0 -> this.sType = "Food";
            case 1 -> this.sType = "Drink";
        }
    }

    public void InputAmount() {
        System.out.print("Amount: ");
        this.iAmount = Main.InputNumber_Int();
    }

    public void InputPrice() {
        System.out.print("Price (VND): ");
        this.dPrice = Main.InputNumber_Double();
    }

    // output
    public int FindPadRight() {
        int max = 1;
        while (this.sID.length() > max
                || this.sName.length() > max
                || this.sType.length() > max
                || String.valueOf(this.iAmount).length() > max
                || String.valueOf(this.dPrice).length() > max
        ) {
            max++;
        }
        return max;
    }

    static public int PadRightMax() {
        int len = Cafe.getlServices().get(0).FindPadRight();
        for (int i = 1; i < Cafe.getlServices().size(); i++) {
            if (Cafe.getlServices().get(i).FindPadRight() > len)
                len = Cafe.getlServices().get(i).FindPadRight();
        }
        return len + 10;
    }

    public void Output() {
        System.out.println("\t" + Main.padRight(this.sID, PadRightMax()) +
                Main.padRight(this.sName, PadRightMax()) +
                Main.padRight(this.sType, PadRightMax()) +
                Main.padRight(String.valueOf(this.iAmount), PadRightMax()) +
                Main.padRight(String.valueOf(this.dPrice), PadRightMax()));
    }

    static public void OutputFields() {
        System.out.println("\n\t" + Main.padRight("[ID]", PadRightMax()) +
                Main.padRight("[NAME]", PadRightMax()) +
                Main.padRight("[TYPE]", PadRightMax()) +
                Main.padRight("[AMOUNT]", PadRightMax()) +
                "[PRICE]");
    }

    public void Input() {
        if (Cafe.getlServices().size() == 0)
            this.sID += 0;
        else
            this.sID += String.valueOf(FindMaxServiceID() + 1);
        System.out.println("Current Service ID: " + this.sID);
        InputName();
        InputType();
        InputAmount();
        InputPrice();
    }

    static public int FindMaxServiceID() {
        int id = 0;
        for (int i = 0; i < Cafe.getlServices().size(); i++) {
            if (Integer.parseInt(Cafe.getlServices().get(i).getsID().substring(2)) > id)
                id = Integer.parseInt(Cafe.getlServices().get(i).getsID().substring(2));
        }
        return id;
    }

    //Methods SORT
    static public void SortID() {
        for (int i = 0; i < Cafe.getlServices().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlServices().size(); j++) {
                if (Cafe.getlServices().get(i).getsID().compareTo(Cafe.getlServices().get(j).getsID()) > 0) {
                    Service tmp = Cafe.getlServices().get(i);
                    Cafe.getlServices().set(i, Cafe.getlServices().get(j));
                    Cafe.getlServices().set(j, tmp);
                }
            }
        }
    }

    static public void SortType() {
        for (int i = 0; i < Cafe.getlServices().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlServices().size(); j++) {
                if (Cafe.getlServices().get(i).getsType().compareTo(Cafe.getlServices().get(j).getsType()) > 0) {
                    Service tmp = Cafe.getlServices().get(i);
                    Cafe.getlServices().set(i, Cafe.getlServices().get(j));
                    Cafe.getlServices().set(j, tmp);
                }
            }
        }
    }

    static public void SortName() {
        for (int i = 0; i < Cafe.getlServices().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlServices().size(); j++) {
                if (Cafe.getlServices().get(i).getsName().compareTo(Cafe.getlServices().get(j).getsName()) > 0) {
                    Service tmp = Cafe.getlServices().get(i);
                    Cafe.getlServices().set(i, Cafe.getlServices().get(j));
                    Cafe.getlServices().set(j, tmp);
                }
            }
        }
    }

    static public void SortAmount() {
        for (int i = 0; i < Cafe.getlServices().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlServices().size(); j++) {
                if (Cafe.getlServices().get(i).getiAmount() > Cafe.getlServices().get(j).getiAmount()) {
                    Service tmp = Cafe.getlServices().get(i);
                    Cafe.getlServices().set(i, Cafe.getlServices().get(j));
                    Cafe.getlServices().set(j, tmp);
                }
            }
        }
    }

    static public void SortPrice() {
        for (int i = 0; i < Cafe.getlServices().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlServices().size(); j++) {
                if (Cafe.getlServices().get(i).getdPrice() > Cafe.getlServices().get(j).getdPrice()) {
                    Service tmp = Cafe.getlServices().get(i);
                    Cafe.getlServices().set(i, Cafe.getlServices().get(j));
                    Cafe.getlServices().set(j, tmp);
                }
            }
        }
    }

    //Login_Methods
    static public int SearchService(String id) {
        id = id.toLowerCase();
        for (int i = 0; i < Cafe.getlServices().size(); i++) {
            if (Cafe.getlServices().get(i).getsID().toLowerCase().equals(id)) {
                return i;
            }
        }
        return -1;
    }

    static public void AddNewService() {
        System.out.println("\n\t[ADDING NEW SERVICE]\n");
        Service sv = new Service();
        sv.Input();
        Cafe.getlServices().add(sv);
    }

    static public int SelectService() {
        String id;
        int pos;
        do {
            try {
                Scanner sc = new Scanner(System.in);
                System.out.print(" => Now Enter Service ID: ");
                id = sc.nextLine();
                pos = SearchService(id);
                if (pos == -1)
                    continue;
                else
                    return pos;
            } catch (Exception e) {
                continue;
            }
        } while (true);

    }

    static public void ResetServiceID() {
        for (int i = 0; i < Cafe.getlServices().size(); i++) {
            Cafe.getlServices().get(i).setsID("SV" + i);
        }
    }

    //Methods FILES
    static public void WriteDataService() throws IOException {
        String[] wt = new String[Cafe.getlServices().size()];
        for (int i = 0; i < Cafe.getlServices().size(); i++) {
            wt[i] = Cafe.getlServices().get(i).getsID() + ";"
                    + Cafe.getlServices().get(i).getsName() + ";"
                    + Cafe.getlServices().get(i).getsType() + ";"
                    + Cafe.getlServices().get(i).getiAmount() + ";"
                    + Cafe.getlServices().get(i).getdPrice();
        }


        File file = new File("Service.txt");
        OutputStream outputStream = new FileOutputStream(file);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputStream);
        for (String item : wt) {
            outputStreamWriter.write(item);
            // Dùng để xuống hàng
            outputStreamWriter.write("\n");
        }
        outputStreamWriter.flush();
    }

    static public void ReadDataService() throws IOException {
        File file = new File("Service.txt");
        InputStream inputStream = new FileInputStream(file);
        InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
        BufferedReader reader = new BufferedReader(inputStreamReader);

        ArrayList<String> a = new ArrayList<>();
        String line;
        while ((line = reader.readLine()) != null) {
            a.add(line);
        }

        for (String s : a) {
            String[] b = s.split(";");
            Service sv = new Service(b[0], b[1], b[2], Integer.parseInt(b[3]), Double.parseDouble(b[4]));
            Cafe.getlServices().add(sv);
        }
    }


}
